### ISSUE TITLE
S4070: False positive with bitwise complement operator

### ISSUE THREAD

**Issue by Treer** [SonarSource/sonar-dotnet#3117]
### Description

The current implementation of the rule raises violation of S4070 (mistaking it for a non-flags enum) for the following code:

    [Flags]
    public enum Permissions
    {
        Create = 1,
        Read   = 2,
        Update = 4,
        Delete = 8,

        None   = 0,
        All    = ~None
    }

If you remove `All = ~None`, the violation goes away.

Or:

    [Flags]
    public enum Foo2
    {
        Developer =  1,
        Tester    =  2,
        Admin     =  4,
        SuperUser =  8,
        User      = 16,
        Auditor   = 32,

        Internal  = Developer | Tester,
        External  = ~Internal  
    }

### Related information

SonarLint 4.17.0.14702 for VisualStudio 2019

**Comment by Corniel** [comment]
IMHO this warning is correct: If we just take the first example.
`All` should have the value  `0xF`, but with your construction has the value `0xFFFFFFFF`.

**Comment by Treer** [comment]
These constructions work correctly with HasFlag(), and *stay correct* when someone adds more values to the enumeration.

Moreover, if you don't want enum 0xFFFFFFFF to represent all flags, then S4070 is the wrong warning message for that.

Twos-compliment means it's easy for a rule to handle the case: The rule already handle numbers up to the sum of the enums, e.g. enums of up to 15 are allowed if you have enums for 1, 2, 4, and 8. Supporting bitwise complement would mean also allowing -1 to -16.

**Comment by Corniel** [comment]
@Treer That is not true:

``` C#

var permissions = (Permissions)15;

if (permissions.HassFlag(Permissions.All)
{
    // should we end up here?!
}
```

The point is, that given your approach, I will not end up within my if statement, but if logically, I should. It is confusing. In other words. Your All contains bit flags that not have been defined.

**Comment by andrei-epure-sonarsource** [comment]
hi @Treer , I agree with @Corniel . Even if you may be using your enum judiciously, in general it could bring lots of confusion.

Also, the [Enum Design](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/enum) guideline from Microsoft says

✔️ DO use powers of two for the flag enum values so they can be freely combined using the bitwise OR operation.

✔️ CONSIDER providing special enum values for commonly used combinations of flags.

`All` is neither a power of two, neither a combination of flags.

```
    [Flags]
    public enum Permissions
    {
        Create = 1,
        Read = 2,
        Update = 4,
        Delete = 8,

        None = 0,
        All = ~None
    }
    class Program
    {

        static void Main(string[] args)
        {

            var allPermisions = Permissions.Create | Permissions.Read | Permissions.Update | Permissions.Delete;

            if (allPermisions.HasFlag(Permissions.All))
            {
                Console.WriteLine("Has flag");
            }
            else
            {
                Console.WriteLine("Does not have all flag"); // --> even if it has all permissions, it doesn't work
            }

            if (allPermisions.HasFlag(allPermisions))
            {
                Console.WriteLine("Has flag"); // --> it works!
            }
            else
            {
                Console.WriteLine("Does not have all flag");
            }
        }
    }
```
